cpp/src/arrow/compute/kernels/scalar_fill_null.cc

```diff
@@ -0,0 +1,223 @@
+  ctx->SetStatus(visitor.GetResult(&result));
+  return result;
+}
+
+struct ScalarFillVisitor {
+  KernelContext* ctx;
+  const ArrayData& data;
+  Datum* out;
+
+  ScalarFillVisitor(KernelContext* ctx, const ArrayData& data, Datum* out)
+      : ctx(ctx), data(data), out(out) {}
+
+  Status Visit(const DataType&) {
+    ArrayData* out_arr = out->mutable_array();
+    *out_arr = data;
+    return Status::OK();
+  }
+
+  Status Visit(const BooleanType&) {
+    const auto& state = checked_cast<const FillNullState<BooleanType>&>(*ctx->state());
+    bool value = UnboxScalar<BooleanType>::Unbox(*state.fill_value);
+    ArrayData* out_arr = out->mutable_array();
+    FirstTimeBitmapWriter bit_writer(out_arr->buffers[1]->mutable_data(), out_arr->offset,
+                                     out_arr->length);
+    FirstTimeBitmapWriter bit_writer_validity(out_arr->buffers[0]->mutable_data(),
+                                              out_arr->offset, out_arr->length);
+    if (data.null_count != 0) {
+      BitmapReader bit_reader(data.buffers[1]->data(), data.offset, data.length);
+      BitmapReader bit_reader_validity(data.buffers[0]->data(), data.offset, data.length);
+      for (int64_t i = 0; i < data.length; i++) {
+        if (bit_reader_validity.IsNotSet()) {
+          if (value == true) {
+            bit_writer.Set();
+          } else {
+            bit_writer.Clear();
+          }
+          bit_writer_validity.Set();
+        } else {
+          if (bit_reader.IsSet()) {
+            bit_writer.Set();
+          } else {
+            bit_writer.Clear();
+          }
+          bit_writer_validity.Set();
+        }
+        bit_reader.Next();
+        bit_writer.Next();
+        bit_reader_validity.Next();
+        bit_writer_validity.Next();
+      }
+      bit_writer_validity.Finish();
+      bit_writer.Finish();
+    } else {
+      *out_arr = data;
+    }
+    return Status::OK();
+  }
+
+  template <typename Type>
+  enable_if_supports_fill_null<Type, Status> Visit(const Type&) {
+    using T = typename GetViewType<Type>::T;
+    const auto& state = checked_cast<const FillNullState<Type>&>(*ctx->state());
+    T value = UnboxScalar<Type>::Unbox(*state.fill_value);
+    const T* in_data = data.GetValues<T>(1);
+    ArrayData* out_arr = out->mutable_array();
+    auto out_data = out_arr->GetMutableValues<T>(1);
+
+    if (data.null_count != 0) {
+      BitmapReader bit_reader(data.buffers[0]->data(), data.offset, data.length);
+      for (int64_t i = 0; i < data.length; i++) {
+        if (bit_reader.IsNotSet()) {
+          out_data[i] = value;
+        } else {
+          out_data[i] = static_cast<T>(in_data[i]);
+        }
+        bit_reader.Next();
+      }
+      BitUtil::SetBitsTo(out_arr->buffers[0]->mutable_data(), out_arr->offset,
+                         out_arr->length, true);
+    } else {
+      *out_arr = data;
+    }
+    return Status::OK();
+  }
+
+  Status Execute() { return VisitTypeInline(*data.type, this); }
```

Comment:
This will need to be reworked to use the `PhysicalType` attributes so that we don't generate functionally identical binary code e.g. for Int64Type / UInt64Type / TimestampType / Date64Type / DurationType (which all use an 8-byte fixed width C type value)

Reply:
Updated